src/main/java/com/timgroup/statsd/StatsDAggregator.java

```diff
@@ -0,0 +1,149 @@
+    protected final Timer scheduler = new Timer(AGGREGATOR_THREAD_NAME, true);
+
+    protected final int shardGranularity;
+    protected final long flushInterval;
+
+    private final StatsDProcessor processor;
+
+    private Telemetry telemetry;
+
+    private class FlushTask extends TimerTask {
+        @Override
+        public void run() {
+            flush();
+        }
+    }
+
+    /**
+     * StatsDAggregtor constructor.
+     *
+     * @param processor     the message processor, aggregated messages will be queued in the high priority queue.
+     * @param shards        number of shards for the aggregation map.
+     * @param flushInterval flush interval in miliseconds, 0 disables message aggregation.
+     *
+     * */
+    public StatsDAggregator(final StatsDProcessor processor, final int shards, final long flushInterval) {
+        this.processor = processor;
+        this.flushInterval = flushInterval;
+        this.shardGranularity = shards;
+        this.aggregateMetrics = new ArrayList<>(shards);
+
+        for (int i = 0 ; i < this.shardGranularity ; i++) {
+            this.aggregateMetrics.add(i, new HashMap<Message, Message>());
+        }
+    }
+
+    /**
+     * Start the aggregator flushing scheduler.
+     *
+     * */
+    public void start() {
+        if (flushInterval > 0) {
+            // snapshot of processor telemetry - avoid volatile reference to harness CPU cache
+            // caller responsible of setting telemetry before starting
+            telemetry = processor.getTelemetry();
+            scheduler.scheduleAtFixedRate(new FlushTask(), flushInterval, flushInterval);
+        }
+    }
+
+    /**
+     * Stop the aggregator flushing scheduler.
+     *
+     * */
+    public void stop() {
+        if (flushInterval > 0) {
+            scheduler.cancel();
+        }
+    }
+
+    public boolean isTypeAggregate(Message.Type type) {
+        return aggregateSet.contains(type);
+    }
+
+    /**
+     * Aggregate a message if possible.
+     *
+     * @param message the dogstatsd Message we wish to aggregate.
+     * @return        a boolean reflecting if the message was aggregated.
+     *
+     * */
+    public boolean aggregateMessage(Message message) {
+        if (flushInterval == 0 || !isTypeAggregate(message.getType()) || message.getDone()) {
+            return false;
+        }
+
+
+        int hash = message.hashCode();
+        int bucket = Math.abs(hash % this.shardGranularity);
+        Map<Message, Message> map = aggregateMetrics.get(bucket);
+
+        synchronized (map) {
+            // For now let's just put the message in the map
+            if (!map.containsKey(message)) {
+                map.put(message, message);
+            } else {
+                Message msg = map.get(message);
+                msg.aggregate(message);
+                if (telemetry != null) {
+                    telemetry.incrAggregatedContexts(1);
+                }
```

Comment:
Shouldn't this be in the if rather than the else ? Sampling the same context over and over should still result in 1 aggregated context.

Reply:
OK, so maybe I thought this meant a different thing. I thought this metric meant the number of total messages we were aggregating per telemetry flush interval. I guess in your mind this means the number of different contexts we are aggregating. I actually think both metrics are useful so I might create another metric. But, that makes sense, I'll move this to the upper block.

Reply:
Yes it's the number of contexts store by the aggregator (very useful to troubleshoot future memory usage question). In Go we update a variable each time we flush. So when the telemetry is flushed it sends the number of contexts hold by the aggregator during its last flush.

For the number of total messages we were aggregating, once we have histogram and sketches aggregated it will be equal to `datadog.dogstatsd.client.metrics`. A useful telemetry would be the number of context/sample per metrics type. To be added for sure in the future (maybe in that `debug` mode we talked about)